src/processors/transform.rs

```diff
@@ -0,0 +1,293 @@
+    #[inline]
+    pub fn parent(&self) -> Option<Entity> { self.parent }
+    #[inline]
+    pub fn set_pos(&mut self, pos: [f32; 3]) {
+        self.pos = pos;
+        self.dirty = true;
+    }
+    #[inline]
+    pub fn set_rot(&mut self, rot: [f32; 4]) {
+        self.rot = rot;
+        self.dirty = true;
+    }
+    #[inline]
+    pub fn set_scale(&mut self, scale: [f32; 3]) {
+        self.scale = scale;
+        self.dirty = true;
+    }
+    #[inline]
+    pub fn set_parent(&mut self, parent: Option<Entity>) {
+        self.parent = parent;
+        self.dirty = true;
+    }
+
+    #[inline]
+    pub fn matrix(&self) -> [[f32; 4]; 4] {
+        let quat: Matrix3<f32> = Quaternion::from(self.rot).into();
+        let mut matrix: Matrix4<f32> = (&quat * Matrix3::new(self.scale[0], 0.0, 0.0,
+                                                             0.0, self.scale[1], 0.0,
+                                                             0.0, 0.0, self.scale[2])).into();
+        matrix.w = Vector3::from(self.pos).extend(1.0f32);
+        matrix.into()
+    }
+}
+
+impl Default for LocalTransform {
+    fn default() -> Self {
+        LocalTransform {
+            pos: [0.0, 0.0, 0.0],
+            rot: [0.0, 0.0, 0.0, 1.0],
+            scale: [1.0, 1.0, 1.0],
+            parent: None,
+            dirty: true,
+        }
+    }
+}
+
+impl Component for LocalTransform {
+    type Storage = VecStorage<LocalTransform>;
+}
+
+/// Absolute transformation (transformed from origin), should be used for rendering position and orientation.
+#[derive(Debug, Copy, Clone)]
+pub struct Transform(pub [[f32; 4]; 4]);
+impl Transform {
+    pub fn identity() -> Self {
+        Transform(
+            [
+                [1.0, 0.0, 0.0, 0.0],
+                [0.0, 1.0, 0.0, 0.0],
+                [0.0, 0.0, 1.0, 0.0],
+                [0.0, 0.0, 0.0, 1.0],
+            ]
+        )
+    }
+}
+
+impl Component for Transform {
+    type Storage = VecStorage<Transform>;
+}
+
+/// Initialization component, added to entity with a `LocalTransform` component after the first update.
+#[derive(Default, Copy, Clone)]
+pub struct Init;
+impl Component for Init {
+    type Storage = NullStorage<Init>;
+}
+
+/// Transformation processor.
+/// Handles updating `Transform` components based on the `LocalTransform` component and parents.
+pub struct TransformProcessor {
+    indices: HashMap<Entity, usize>,
+    sorted: VecDeque<(Entity, Option<Entity>)>,
+}
+
+impl TransformProcessor {
+    pub fn new() -> TransformProcessor {
+        TransformProcessor {
+            indices: HashMap::new(),
+            sorted: VecDeque::new(),
+        }
+    }
+}
+
+impl Processor<Arc<Mutex<Context>>> for TransformProcessor {
+    fn run(&mut self, arg: RunArg, _: Arc<Mutex<Context>>) {
+        let (entities, mut locals, mut globals, mut init) = arg.fetch(|w| {
+            (w.entities(), w.write::<LocalTransform>(), w.write::<Transform>(), w.write::<Init>())

```

Comment:
I don't think locking the `LocalTransform` for write is a good idea. I assume that's in order to change the `dirty` flag? If that so, `dirty` should just be wrapped into a `Cell`.